[Spark-12732][ML] bug fix in linear regression train

iyounus opened the pull request:
Fixed the bug in linear regression train for the case when the target variable is constant. The two cases for `fitIntercept=true` or `fitIntercept=false` should be treated differently.

iyounus committed:
bug fix in linear regrssion for contstant label, but the test still fails.

iyounus committed:
fixed the test case for constant label

SparkQA commented:
**[Test build #49163 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49163/consoleFull)** for PR 10702 at commit [`944be71`](https://github.com/apache/spark/commit/944be7143880bc307d8538ae1e354ea2c184787b).

srowen commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -219,33 +219,41 @@ class LinearRegression @Since("1.3.0") (@Since("1.3.0") override val uid: String
 
     val yMean = ySummarizer.mean(0)
-    val yStd = math.sqrt(ySummarizer.variance(0))
+    var yStd = math.sqrt(ySummarizer.variance(0))
```
Can you put all this logic up here?
```
val rawYStd = math.sqrt(ySummarizer.variance(0))
val yStd = if (rawYStd > 0.0 || fitIntercept) {
  rawYStd
} else {
  logWarning(...)
  1.0
}

dbtsai replied:
Also, when `rawYStd == 0.0`, `standardization == true` and `regParam != 0.0`, the problem will be ill-defined. We may want to throw an exception.

srowen replied:
@iyounus what do you think of these comments?

iyounus replied:
@srowen I agree with these comments. I'm working on `WeightedLeastSquares` (https://github.com/apache/spark/pull/10274), and there are some commonalities between these two issues. Once I've completed all tests for `WeightedLeastSquares`, I'll compete this issue.

Some of the tests for `LinearRegression` use both `normal` and `l-bfgs` solvers. It would nice if `WeightedLeastSquares` issue is merged so that I can write tests in similar way.

SparkQA commented:
**[Test build #49163 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49163/consoleFull)** for PR 10702 at commit [`944be71`](https://github.com/apache/spark/commit/944be7143880bc307d8538ae1e354ea2c184787b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49163/
Test PASSed.

mengxr commented:
@dbtsai Do you have time to make a pass?

iyounus commented:
@mengxr I haven't implemented the changes suggested by @dbtsai and @srowen yet. It think the solution I proposed to this issue may not be very suitable. I'll make some updates to this PR either today or tomorrow.

SparkQA commented:
**[Test build #49816 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49816/consoleFull)** for PR 10702 at commit [`23ce5f3`](https://github.com/apache/spark/commit/23ce5f3eef773b9fec39e6fccd863760f594c3f6).

dbtsai commented:
@mengxr This PR is also on my radar. Working on another PR now, once @iyounus is ready, I will work on this.

SparkQA commented:
**[Test build #49816 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49816/consoleFull)** for PR 10702 at commit [`23ce5f3`](https://github.com/apache/spark/commit/23ce5f3eef773b9fec39e6fccd863760f594c3f6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49816/
Test PASSed.

iyounus committed:
added exception for ill-defined case of constant label with regParam=0, standardization=true and fitIntercept=false. added test for this case. Modified existing test for constant label.

SparkQA commented:
**[Test build #49889 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49889/consoleFull)** for PR 10702 at commit [`5803bd1`](https://github.com/apache/spark/commit/5803bd146307d6fd42070574da0a490646d866b2).

iyounus commented:
I've added an exception for the case when label is constant and `standardization == true` and `regParam != 0.0`. Also added test for this case.

I cannot test the case when `standardizeLabel = false` with non-zero `regParam` against analytic normal equation solution, because the analytic normal equation gives me quite different results.

Even for the case when `regParam=0` and `fitIntercept=false`, the results from normal equation are slightly different from `glmnet` and `spark`. Please see my detailed comment regarding this comparison at https://github.com/apache/spark/pull/10274.

SparkQA commented:
**[Test build #49889 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/49889/consoleFull)** for PR 10702 at commit [`5803bd1`](https://github.com/apache/spark/commit/5803bd146307d6fd42070574da0a490646d866b2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/49889/
Test PASSed.

dbtsai commented:
@iyounus `standardizeLabel = false/ture` with non-zero `regParam`, let's throw the exception. I explained the mismatch against the analytic normal equation in the other PR. 

Thanks.

dbtsai commented on mllib/src/test/scala/org/apache/spark/ml/regression/LinearRegressionSuite.scala:
```diff
@@ -558,6 +575,47 @@ class LinearRegressionSuite
+
+  test("regularized linear regression through origin with constant label") {
+    // The problem is ill-defined if fitIntercept=false, regParam is non-zero and \
+    // standardization=true. An exception is thrown in this case.
```
When `standardization=false`, the problem is still ill-defined since GLMNET always standardizes the labels. That's why you see it in the analytical solution. Let's throw exception when `fitIntercept=false` and `regParam != 0.0`.

dbtsai commented on mllib/src/test/scala/org/apache/spark/ml/regression/LinearRegressionSuite.scala:
```diff
@@ -558,6 +575,47 @@ class LinearRegressionSuite
+  }
+
+  test("regularized linear regression through origin with constant label") {
+    // The problem is ill-defined if fitIntercept=false, regParam is non-zero and \
```
Remove `\` in the end of line.

dbtsai replied:
seems you forget to remove `\` :)

dbtsai commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -219,33 +219,43 @@ class LinearRegression @Since("1.3.0") (@Since("1.3.0") override val uid: String
+          Array(0D))
+        return copyValues(model.setSummary(trainingSummary))
+      } else {
+        require(!($(regParam) > 0.0 && $(standardization)),
```
remove `&& $(standardization)`

dbtsai replied:
Just 
` require($(regParam) != 0.0)`

dbtsai replied:
also change the message.

iyounus replied:
I can change this. But, the behaviour of WeightedLeastSquares should also be the same. Should I also make changes to WeightedLeastSquares?

dbtsai replied:
Your PR for `WeightedLeastSquares` is doing the right behavior, and consistent to `require($(regParam) != 0.0)`. The problem only happens when `standardizationLabel = true`, and in your `WeightedLeastSquares`, you already throw exception in this case. 

GLMNET will do `standardizationLabel = true` even `standardization == false`, and that's why you see inconsistent solution between GLMNET and your analytical solution.

dbtsai commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -219,33 +219,43 @@ class LinearRegression @Since("1.3.0") (@Since("1.3.0") override val uid: String
 
+    // if y is constant (rawYStd is zero), then y cannot be scaled. In this case
+    // setting yStd=1.0 ensures that y is not scaled anymore in l-bfgs algorithm.
+    val yStd = if (rawYStd > 0) rawYStd else 1.0
```
I will do
```scala
val yStd = if (rawYStd > 0) rawYStd else if(yMean != 0.0) math.abs(yMean) else 1.0
```

dbtsai replied:
Actually, in the case of `yMean == 0.0 && yStd == 0.0`, the coefficients will be all zeros as well even `fitIntercept  == false`. This is rare case, so we can let model training to figure out. But if you want to handle this explicitly, it's great.

iyounus replied:
Its not clear to me why would you set yStd = abs(yMean) if label is constant.

dbtsai replied:
This will help the optimization if `yMean` is very large. Kind of we are normalizing the label instead of standardizing the label.

iyounus committed:
fixed the case with yStd==0, regParam!=0 and standardization=true/false. Also, if yMean==0 and yStd==0, no training is done.

SparkQA commented:
**[Test build #50450 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50450/consoleFull)** for PR 10702 at commit [`e83b822`](https://github.com/apache/spark/commit/e83b8223846cc41942469fc4b78e9f0500239e0f).

SparkQA commented:
**[Test build #50450 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50450/consoleFull)** for PR 10702 at commit [`e83b822`](https://github.com/apache/spark/commit/e83b8223846cc41942469fc4b78e9f0500239e0f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50450/
Test PASSed.

iyounus commented:
I've completed this PR. I think all the tests are there. Here, I'm going to document a couple of minor issues just for future reference.

__Issue 1__
For the case when `yStd = 0` and `fitIntercept = false`, we've four possibilities (`reParam: zero/non-zero` and `standardization: true/false`). Using `WeightedLeastSquares` (`normal` solver), I _can_ get the following results:
```
# data used for the following results
val df = sc.parallelize(Seq(
  (17.0, Vectors.dense(0.0, 5.0)),
  (17.0, Vectors.dense(1.0, 7.0)),
  (17.0, Vectors.dense(2.0, 11.0)),
  (17.0, Vectors.dense(3.0, 13.0))
), 2).toDF("label", "features")
```

```
# coefficients obtained from WeightedLeastSquares
(1) reg: 0.0, standardization: false
--------> 0.0 [-9.508474576271158,3.457627118644062]

(2) reg: 0.0, standardization: true
--------> 0.0 [-9.508474576271158,3.457627118644062]

(3) reg: 0.1, standardization: false
--------> 0.0 [-7.134240246406588,3.010780287474336]

(4) reg: 0.1, standardization: true
--------> 0.0 [-5.730337078651679,2.7219101123595495]
```
This is with `L2` regularization, and ignoring standardization of the label for the case (4). For the case (4), we throw an error because this is ill-defined, so the user never sees these results.

For case (3), even though the standardization is `false`, the label is still standardized because the `standardizeLable` is hardwired to be `true` when calling `WeightedLeastSquares` within `LinearRegression` class. Therefore, an error is thrown in this case too. Which, in my opinion, is not right thing to do because the analytical solution does exist for this case.

__Issue 2__
Again, for the case when `yStd = 0` and `fitIntercept = false`, I can get the following results using `l-bfgs` solver:

```
(1) reg: 0.0, standardization: false
--------> 0.0 [-9.508474576271176,3.4576271186440652]

(2) reg: 0.0, standardization: true
--------> 0.0 [-9.508474576271176,3.4576271186440652]

(3) reg: 0.1, standardization: false
--------> 0.0 [-9.327614273741196,3.423618722197146]

(4) reg: 0.1, standardization: true
--------> 0.0 [-9.08129403505256,3.374915377479131]
```

Here, results (1) and (2) are identical to what we get from `WeightedLeastSquares` as expected. Case (4) is ill-defined and we throw an error.

Now, for case (3), the numerical values are different as compared to `WeightedLeastSquares`. This is because we standardize label using `yMean`. Otherwise, the values obtained from `l-bfgs` are identical to `WeightedLeastSquares`. Note that the user will not see these values because an error is thrown for this case instead.

__Issue 3__
The normal equation with regression (Ridge Regression), gives significantly different results as compared to case (3) above. Here is my R code with results:
```
ridge_regression <- function(A, b, lambda, intercept=TRUE){
    if (intercept) {
        A = cbind(rep(1.0, length(b)), A)
        I = diag(ncol(A))
        I[1,1] = 0.0
    } else {
        I = diag(ncol(A))
    }
    R = chol( t(A) %*% A + lambda*I )
    z = solve(t(R), t(A) %*% b)
    w = solve(R, z)
    return(w)
}
A <- matrix(c(0, 1, 2, 3, 5, 7, 11, 13), 4, 2)
b <- c(17, 17, 17, 17)
df <- as.data.frame(cbind(A, b))

ridge_regression(A, b, 0.1, intercept = FALSE)

[1,] -8.783272
[2,]  3.321237
```
In my opinion, when `standardization=flase`, the results from `normal` solver must match this. Even though the user doesn't see this case, it gives me less confidence in the implementation of normal equation, because it doesn't match this simple case. I also wrote about this at https://github.com/apache/spark/pull/10274.

dbtsai commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -219,33 +219,44 @@ class LinearRegression @Since("1.3.0") (@Since("1.3.0") override val uid: String
 
+    // if y is constant (rawYStd is zero), then y cannot be scaled. In this case
+    // setting yStd=1.0 ensures that y is not scaled anymore in l-bfgs algorithm.
+    val yStd = if (rawYStd > 0) rawYStd else if (yMean != 0.0) math.abs(yMean) else 1.0
```
`val yStd = if (rawYStd > 0) rawYStd else math.abs(yMean)' since you already check the condition before.

iyounus committed:
forgot to add this file with previous commit

dbtsai commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -219,33 +219,44 @@ class LinearRegression @Since("1.3.0") (@Since("1.3.0") override val uid: String
+        // zero coefficient; as a result, training is not needed.
+        // Also, if yMean==0 and rawYStd==0, all the coefficients are zero regardless of
+        // the fitIntercept
+        logWarning(s"The standard deviation of the label is zero, so the coefficients will be " +
```
Maybe you want to update the warning message for the second situation as well.

dbtsai commented on mllib/src/test/scala/org/apache/spark/ml/regression/LinearRegressionSuite.scala:
```diff
@@ -558,6 +575,47 @@ class LinearRegressionSuite
+          .fit(datasetWithWeightConstantLabel)
+        val actual = Vectors.dense(model.intercept, model.coefficients(0), model.coefficients(1))
+        assert(actual ~== expected(idx) absTol 1e-4)
+        idx += 1
```
When `fitInercept = true`, check the size of loss history is zero. (since the solution is returned without any optimization.) 

Will be nice to add one small test that `labelStd = 0` and `labelMean = 0` when `fitInercept = false`

iyounus replied:
I'm not sure how to _check the size of lost history_. Could you please point me to some example?

dbtsai replied:
In `LinearRegressionTrainingSummary`, you can get it from `objectiveHistory`.

iyounus replied:
I've added tests as suggested.

dbtsai commented:
LGTM except minor comments. Thanks.

SparkQA commented:
**[Test build #50454 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50454/consoleFull)** for PR 10702 at commit [`0b16353`](https://github.com/apache/spark/commit/0b16353a6777799fbcdea50e0f0213a405934447).

iyounus committed:
minor changes.

SparkQA commented:
**[Test build #50455 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50455/consoleFull)** for PR 10702 at commit [`c0744d8`](https://github.com/apache/spark/commit/c0744d8a3c08756546925c9f82274f50d1d4affd).

dbtsai commented:
Commenting on your issues. 

Issue 1:
With `WeightedLeastSquares`, we have option to standardize the label and features separately. As a result, if the label is not standardized, even `yStd == 0`, the problem can be solved. 

As a result, in your case 4, when label is not standardized, and the features are standardized, this is not defined, so the users should get the result.

For case 3, can you elaborate why analytical solution exists even the label is standardized? 

Issue 2: 

In my opinion, even case 1, and case 2 are ill-defined since in GLMNET, the label is standardized by default, and GLMNET will not return any result at all. It just happens that without regularization, with/without standardization on labels will not change the solution, so we just treat them as if we don't standardize the label. This can explain your case 3.

Issue 3:

I think this is because your normal equation solver doesn't standardize the label, so the discrepancies occur.

SparkQA commented:
**[Test build #50454 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50454/consoleFull)** for PR 10702 at commit [`0b16353`](https://github.com/apache/spark/commit/0b16353a6777799fbcdea50e0f0213a405934447).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50454/
Test PASSed.

SparkQA commented:
**[Test build #50455 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50455/consoleFull)** for PR 10702 at commit [`c0744d8`](https://github.com/apache/spark/commit/c0744d8a3c08756546925c9f82274f50d1d4affd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50455/
Test PASSed.

iyounus commented:
For the case (3), I'm assuming that the label and features are not standardized. So, in that case, the solution exists. Here is my perspective on this.

The normal equation `X^T X \beta = X^T y` has a unique solution if matrix `X` is full rank. If its not full rank, the `X^T X` becomes singular and hence not invertible. With regularization, the equation  `(X^T X + \lambda I) \beta = X^T y` still has unique solution if the matrix `(X^T X + \lambda I)` is invertible. This is true even if elements of `y` are constant. For the sample data I'm using above, I can solve this equation by hand and obtain `\beta` (with and without intercept). If I'm using any software, it must reproduce these results. Note that standardization is completely independent operation. Its not part of normal equation. So, if the user demands no standardization, then the any software should produce the analytical solution.

iyounus committed:
added test for the case when results are produced without training (when label is constant)

SparkQA commented:
**[Test build #50512 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50512/consoleFull)** for PR 10702 at commit [`2480dc1`](https://github.com/apache/spark/commit/2480dc17b064165563fe30bf5e45d2d165c722d9).

SparkQA commented:
**[Test build #50512 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50512/consoleFull)** for PR 10702 at commit [`2480dc1`](https://github.com/apache/spark/commit/2480dc17b064165563fe30bf5e45d2d165c722d9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50512/
Test PASSed.

dbtsai commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -74,7 +74,8 @@ class LinearRegression @Since("1.3.0") (@Since("1.3.0") override val uid: String
   /**
    * Set the regularization parameter.
    * Default is 0.0.
```
All the indentations you just added are off.

dbtsai commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -83,7 +84,8 @@ class LinearRegression @Since("1.3.0") (@Since("1.3.0") override val uid: String
    * Set if we should fit the intercept
    * Default is true.
-   * @group setParam
+    *
```
ditto

dbtsai commented:
For the case (3), I agree with your agreement completely. Can you try your normal equation solution with L2 without any standardization (nonzero ystd data) and see if the result match GLMNET?  If I remember correctly, this will not match since GLMNET internally will standardize labels even one has `standardization = false`. If this is true, when ystd = 0, it doesn't make sense to have different rule, and after all, GLMNET will just return error in this case.

dbtsai commented on mllib/src/main/scala/org/apache/spark/ml/regression/LinearRegression.scala:
```diff
@@ -398,7 +422,8 @@ class LinearRegressionModel private[ml] (
   /**
    * Evaluates the model on a testset.
-   * @param dataset Test dataset to evaluate model on.
+    *
```
ditto

iyounus replied:
I really don't know how this has happened. :)
I've fixed these. All the indentation should be in order now.

iyounus committed:
fixed indentation

SparkQA commented:
**[Test build #50575 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50575/consoleFull)** for PR 10702 at commit [`fd7eb99`](https://github.com/apache/spark/commit/fd7eb99fb7d71e8c799f05a601f46c0001c7debd).

iyounus commented:
GLMNET sets all coefficients to zero if yStd=0 and fitIntercept=false regardless of standardization or regularization. Thats why I cannot compare my normal equation with GLMNET.

SparkQA commented:
**[Test build #50575 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50575/consoleFull)** for PR 10702 at commit [`fd7eb99`](https://github.com/apache/spark/commit/fd7eb99fb7d71e8c799f05a601f46c0001c7debd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50575/
Test PASSed.

dbtsai commented:
I meat comparing the result with your solution when `yStd != 0`, and `regParm != 0`. I suspect that you will get different result since GLMNET one forces to standardize the labels even standardization is off.

iyounus commented:
For `yStd != 0`, and `regParm != 0`, my solution doesn't match with GLMNET. I showed this comparison on this jira https://github.com/apache/spark/pull/10274.

dbtsai commented:
Yes, that's what I meat. Without standardizing the labels, no way to match glmnet, but this makes the problem ill-defined when `yStd == 0`.

dbtsai commented on mllib/src/test/scala/org/apache/spark/ml/regression/LinearRegressionSuite.scala:
```diff
@@ -558,6 +583,86 @@ class LinearRegressionSuite
+          .setSolver("l-bfgs")
+          .fit(datasetWithWeightConstantLabel)
+        if (fitIntercept) {
+          assert(model1.summary.objectiveHistory(0) ~== 0.0 absTol 1e-4)
```
I think this should be `model1.summary.objectiveHistory.length == 0`

dbtsai commented:
LGTM. Merged into master. Thanks.
